### PR TITLE
hdf5: solve cmake findhdf5 issue with hl components

### LINKED ISSUE #44422: Installation issue: OpenFAST+cxx build failure
### Steps to reproduce the issue

```console
git clone https://github.com/spack/spack
cd spack
git checkout v0.22.0
source share/spack/setup-env.sh
spack compiler find
spack install openfast%gcc@8.5.0+cxx
```

### Error message

<details><summary>Error message</summary>

<pre>
1 error found in build log:
     33    -- Found MPI_C: /tmp/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/openmpi-5.0.3-f5gvuwbek3dovg26bry7wkth42mbwii3
           /bin/mpicc (found version "3.1")
     34    -- Found MPI_CXX: /tmp/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/openmpi-5.0.3-f5gvuwbek3dovg26bry7wkth42mbwi
           i3/bin/mpic++ (found version "3.1")
     35    -- Found MPI_Fortran: /tmp/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/openmpi-5.0.3-f5gvuwbek3dovg26bry7wkth42
           mbwii3/bin/mpif90 (found version "3.1")
     36    -- Found MPI: TRUE (found version "3.1")
     37    -- Found LibXml2: /tmp/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/libxml2-2.10.3-dcunwsit6kkxcifdvntmhwzcki4ts
           uk3/lib/libxml2.so (found version "2.10.3")
     38    -- Found ZLIB: /tmp/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/zlib-ng-2.1.6-2cikdarjflejn3zjgvsletvdajgc74lp/
           lib/libz.so (found version "1.3.0")
  >> 39    CMake Error at /tmp/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/cmake-3.27.9-vo3f3v7ppcksymyr6iaqiwfvnc36uhnc/s
           hare/cmake-3.27/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
     40      Could NOT find HDF5 (missing: HDF5_HL_LIBRARIES) (found version "1.14.3")
     41    Call Stack (most recent call first):
     42      /tmp/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/cmake-3.27.9-vo3f3v7ppcksymyr6iaqiwfvnc36uhnc/share/cmake-3.
           27/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
     43      /tmp/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/cmake-3.27.9-vo3f3v7ppcksymyr6iaqiwfvnc36uhnc/share/cmake-3.
           27/Modules/FindHDF5.cmake:1025 (find_package_handle_standard_args)
     44      glue-codes/openfast-cpp/CMakeLists.txt:29 (find_package)

</pre></details>

### Information on your system

* **Spack:** 0.22.0 (5fe93fee1eec46a0750bd340198bffcb92ff9eec)
* **Python:** 3.6.8
* **Platform:** linux-rhel8-cascadelake
* **Concretizer:** clingo

### Additional information

Dear maintainer @jrood-nrel
Building OpenFAST with the variant +cxx fails on v0.22.0 and v0.21.2 and switching to a different gcc version (11.3.0) produces the same result.
Because HDF5 is a dependency brought in with the +cxx variant, i'm not sure if i can rule out the presence of a bug in CMake's FindHDF5.cmake which might cause the HDF5_HL_LIBRARY detection to fail, although it did find the correct HDF5 version, see the (debug.log)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by jrood-nrel**
There is probably a better way to do this, but this how we have been solving this issue for a long time, which was reported here https://github.com/spack/spack/issues/44422.

Closes https://github.com/spack/spack/issues/44422

### CODE REVIEW

**Review by hyoklee — APPROVED** [review]
Thank you for your contribution, @jrood-nrel !

**Review by psakievich — DISMISSED** [review]

**Comment by haampie** [comment]
Sounds primarily like a CMake bug. Is there an upstream issue for it?

**Comment by psakievich** [comment]
@haampie there is indeed: https://gitlab.kitware.com/cmake/cmake/-/issues/23261

2 years and counting

**Comment by jrood-nrel** [comment]
Wow look at me making a CMake issue for it. Totally forgot about it.

**Review by hyoklee — APPROVED** [review]

**Review by eugeneswalker — DISMISSED** [review]
Thanks! 

Does anyone else want to comment before merge?

File: var/spack/repos/builtin/packages/hdf5/package.py [hunk 1]
```diff
@@ -621,6 +621,17 @@ def link_debug_libs(self):
...
+    @run_after("install")
+    def symlink_to_h5hl_wrappers(self):
+        if self.spec.satisfies("+hl"):
+            with working_dir(self.prefix.bin):
+                # CMake's FindHDF5 relies only on h5cc so it doesn't find the HL
+                # component unless it uses h5hlcc so we symlink h5cc to h5hlcc etc
+                os.remove("h5cc")
+                os.remove("h5c++")
+                symlink("h5hlcc", "h5cc")
+                symlink("h5hlc++", "h5c++")
```
Comment (psakievich) [inline var/spack/repos/builtin/packages/hdf5/package.py]:
```suggestion
                symlink_files = {
                                  "h5cc":  "h5lcc",
                                  "h5c++": "h5lc++"
                 }
                 for old, new in symlink_files.items():
                     if os.path.isfile(old):
                         os.remove(old)
                         symlink(new, old)
```
Looks like this is failing the CI pipelines b/c these files aren't always there? This suggestion will hopefully resolve that, but it is likely a version issue that needs to be satisfied along with the `+hl`?

**Review by hyoklee — APPROVED** [review]

**Review by psakievich — APPROVED** [review]

**Comment by jrood-nrel** [comment]
It looks like CMake may finally be fixing this on their end https://gitlab.kitware.com/cmake/cmake/-/issues/23261#note_1634698

**Pull request merged**